---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# microbiome10XVisium

<!-- badges: start -->
<!-- badges: end -->

microbiome10XVisium is part of the *SpatialMicrobe* computational framework, which allows to profile microbial reads from 10XVisium Spatial Gene Expression data. It is used downstream of the snakemake pipeline X. 

## Installation

You can install the development version of microbiome10XVisium from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("bedapub/spatial-microbe")
```

## Basic use case

This is a basic example which shows you how to solve a common problem:

```{r example}
library(microbiome10XVisium)
## basic example code
```

We are going to demonstrate the basic workflow to profile microbial reads in 10X Visium spatial transcriptomics samples with the CRC_16 sample. CRC_16 is a colorectal cancer sample form the Galeano-Ni√±o et al. (Nature 2022) publication. 
Using the **microbiome10XVisium** package requires that the snakemake pipeline was run beforehand.

The steps of the workflow involve:

1) **krakenToMatrix()**: barcode correction, UMI deduplication and resolving the reads at a specific taxonomic level
2) **decontaminate()**: running various decontamination steps and adding the microbiome information as an additional assay to a Seurat object containing the GEX assay
3) Downstream analysis and visualization, such as visualizing the pseudobulk microbiome composition, spatial profiles of certain microbes or co-occurrence of taxa in spots.

### Preparation

In case of using the microbiome10XVisium for the first time, it is required to download a SQLite DB for taxonomic conversion (performed with the *taxonomizr* R package, for more info: https://github.com/sherrillmix/taxonomizr). Run the following code once:

```{r eval=FALSE, include=TRUE}
library(taxonomizr)
taxonomizrDB = "DESIRED/FILEPATH/FOR/SQLite/DB/nameNode.sqlite"
taxonomizr::prepareDatabase(sqlFile=taxonomizr_DB, getAccessions=FALSE)
```

Specify the location of the SQLite DB for downstream use:
```{r}
taxonomizrDB="/projects/site/pred/microbiome/database/taxonomizr_DB/nameNode.sqlite"
```

### krakenToMatrix()

The first step is to convert the output from the bioinformatic pipeline (*_kraken_output.txt file consisting of three columns BC, UMI and taxid) into a taxid-spot matrix at a particular taxonomic level, similar to the gene-spot matrix for host transcriptomics. The function krakenToMatrix() performs this, while also correcting the barcodes (i.e. only selecting valid barcodes present in the 10X Visium whitelist), resolving the taxonomic classifications of the reads to a specific taxonomic level (recommended: genus level) and collapsing reads into UMI counts. 

Running **krakenToMatrix()** with default settings (i.e. tax_level="genus" and counts="umi_counts"): 
```{r}
k2m <- krakenToMatrix(
  filePath=system.file("extdata", "CRC_16", "CRC_16_kraken_output.txt.gz", package="microbiome10XVisium"),
  outDir=system.file("extdata", "CRC_16/", package="microbiome10XVisium"), taxonomizrDB=taxonomizrDB)
```

This returns a list consisting of \$taxid_counts (a dataframe containing the pseudobulk composition of the sample) and \$matrix (the taxid-spot matrix). 
```{r results='asis'}
knitr::kable(head(k2m$taxid_counts, n=5))
```
As we can see, the top 5 taxa detected in this sample are **Fusobacterium**, **Bacteroides**, **Leptotrichia**, **Campylobacter** and **Gemella**.

```{r}
dim(k2m$matrix)
```
All in all, 962 genera were detected in the 4992 spots of the 10X Visium tissue slide.

### decontaminate()

The next step is to decontaminate the taxid-spot matrix. There are multiple options for decontamination and we recommend an iterative approach for each dataset to find the optimal decontamination strategy (i.e. revisiting the decontamination after looking for batch effects in the dataset). The decontamination can be performed on four levels:

1) **removeSingletons**: removes singleton counts (i.e. values of 1 in the taxid-spot matrix), which are probable false-positives from the taxonomic classification process.
2) **removeLikelyContaminants**: removes taxa that are likely contaminants (defined by Poore et al. (Nature 2020)). **removeSpecificTaxa**: removes additional taxa defined by the users that are likely contaminants or batch effects (for example Mycobacterium). Important: this only works at genus level!
3) **selectGastrointestinal**: only keeps taxa that are likely gastrointestinal (oral or fecal) commensals (defined by Schmidt et al. (eLife 2019)). Important: this only work at genus level!
4) **spots**: this selects spots from the 10X Visium slide. The 10X Visium tissue slide consists of 4992 spots, but only some of the spots are covered by tissue (termed "tissueOnly"). The user has the option to only choose tissue-covered spots ("tissueOnly"), tissue-covered spots and additionally spots that are bordering the tissue ("tissuePlusBordering") or to select all spots on the slide ("all"). We propose to run "all" first to get an overview of the microbial signal on the slide, and then to choose either "tissueOnly" or "tissuePlusBordering" (in case of high microbial signal in tissue bordering region) for the final decontaminated version. 

#### all spots
```{r message=FALSE, warning=FALSE}
#all spots
CRC_16_all <- decontaminate(sampleName = "CRC_16",
              object=k2m,
              spacerangerDir = system.file("extdata", "outs/", package="microbiome10XVisium"),
              outDir = system.file("extdata", "CRC_16/", "vignette/", package="microbiome10XVisium"),
              spots="all",
              removeSingletons=TRUE,
              removeLikelyContaminants = TRUE,
              taxonomizrDB=taxonomizrDB)
```

Looking at the spatial microbiome profile after this round of decontamination:
```{r results='asis'}
spatialPlot(CRC_16_all, taxa="all", taxonomizrDB=taxonomizrDB)
```

We can see that most of the signal is located in tissue-covered spots and detect that there is a weird area with high microbial signal on the left, not related to the tissue. Thus we can proceed with only including the tissue-covered spots ("tissueOnly"). 

#### only considering tissue-covered spots

```{r message=FALSE, warning=FALSE}
#all spots
CRC_16_tissueOnly <- decontaminate(sampleName = "CRC_16",
              object=k2m,
              spacerangerDir = system.file("extdata", "outs/", package="microbiome10XVisium"),
              outDir = system.file("extdata", "CRC_16/", "vignette/", package="microbiome10XVisium"),
              spots="tissueOnly",
              removeSingletons=TRUE,
              removeLikelyContaminants = TRUE,
              taxonomizrDB=taxonomizrDB)
```
Through the decontamination process we reduced the total number of taxa detected from 962 to 155.

Looking at the spatial microbiome profile after this round of decontamination:
```{r results='asis'}
spatialPlot(CRC_16_tissueOnly, taxa="all", taxonomizrDB=taxonomizrDB)
```

### Downstream analyis and visualization

To get an overview of what the most abundant microbial taxa are in this sample after decontamination, we can have a look at the $taxid_counts dataframe.
```{r results='asis'}
knitr::kable(head(CRC_16_tissueOnly$taxid_counts))
```

#### Spatial profiles
Looking at the spatial profile of these top 3 taxa:
```{r results='asis'}
spatialPlot(CRC_16_tissueOnly, taxa="Fusobacterium", taxonomizrDB=taxonomizrDB)
spatialPlot(CRC_16_tissueOnly, taxa="Bacteroides", taxonomizrDB=taxonomizrDB)
spatialPlot(CRC_16_tissueOnly, taxa="Leptotrichia", taxonomizrDB=taxonomizrDB)
```

The top three taxa seem to be most prevalent in one tissue region at the bottom of the slide. 

Investigating the spatial composition of the sample in a Shiny plot (only works interactively!)
```{r}
#interactive_spatialPlot(object=CRC_16_tissueOnly)
```

#### Co-occurrence analysis

We can also investigate whether certain taxa co-occur in the same spots.

Looking at the number of taxa that are detected in each spot:
```{r}
spatialPlot(object=CRC_16_tissueOnly, taxa="nTaxa") + ggplot2::ggtitle("nTaxa detected")
```

There are definitely multiple taxa per spot in the spots with high microbial signal!

Creating a co-occurrence network (at different SparCC correlation thresholds):

```{r message=FALSE, warning=FALSE, fig.fullwidth=TRUE}
cooccurrenceNetwork(object=CRC_16_tissueOnly, threshold=0.5) 
```

There seems to be some co-occurrence of the top 5 most abundant taxa in the sample: **Fusobacterium**, **Bacteroides**, **Leptotrichia**, **Campylobacter** and **Gemella**.

#### Pseudobulk analyis
Moving away from the spatial profiles and looking at the pseudobulk relative abundance profiles instead:
```{r fig.fullwidth=TRUE}
pseudoBulkProfile(sampleName="CRC_16", object=CRC_16_tissueOnly) + ggplot2::ggtitle("")
```

### exporting the decontaminated taxid-spot matrix

The decontaminated taxid-spot matrix can be exported as csv file - to be integrated in other analyses (outside of R and the Seurat package).

```{r}
# with taxids as matrix rownames
export_matrix(object=CRC_16_tissueOnly, rownames="taxid", outDir=system.file("extdata", "CRC_16/", package="microbiome10XVisium"), taxonomizrDB=taxonomizrDB)
# with genus names as matrix rownames
export_matrix(object=CRC_16_tissueOnly, rownames="genus", outDir=system.file("extdata", "CRC_16/", package="microbiome10XVisium"), taxonomizrDB=taxonomizrDB)
```

